### ISSUE TITLE
online tracking

### ISSUE THREAD

**Issue by XMraja** [yformer/EfficientTAM#7]
Thanks for the work.
Does this work support online tracking?

**Comment by okideal** [comment]
The overall process and architecture are the same. You can refer to this project. [segment-anything-2-real-time](https://github.com/Gy920/segment-anything-2-real-time)

**Comment by yformer** [comment]
Hi @XMraja, Thanks for your interest in our work!

Yes, it supports online tracking.

**Comment by yformer** [comment]
@okideal, we just released a family of efficient track anything checkpoints. Are you interested in integrating it to your repo?

**Comment by okideal** [comment]
@yformer Glad to see that the checkpoints have been uploaded!! 
Gonna to test the accuracy first and then perform deployment using TensorRT.

**Comment by yformer** [comment]
@okideal Cool. Let me know if you have any questions.

**Comment by okideal** [comment]
@yformer I have integrated TensorRT in C++. So far, the testing results look good, achieving about 20 ms for full process inference on a 4090 GPU (with the efficienttam_s.pt version).

For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (dynamic speed is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?

**Comment by okideal** [comment]
In addition, I noticed that the decoder's output mask has only one, even though the configuration file sets **multimask_output_for_tracking** to true. Is there something I haven’t handled correctly? 
If there is only one output mask, does that mean the optimization features for this project [https://github.com/yangchris11/samurai](samurai) cannot be applied?

**Comment by rodneyML** [comment]
Hi @okideal . I am also interested in integrating Efficient Track Anything with TensorRT and C++. Do you perhaps have a repo where you did that? If you don't mind sharing, how did you build the engine file? Did you convert the whole model into a single ONNX file or did you split it up in multiple stages? Thanks in advance!

@yformer a lot of people I think will benefit if a `torch.export` script is provided to convert this model into a ONNX file (and then ultimately TensorRT). Great work btw, have been keeping my eye on this repo since day 1!

**Comment by okideal** [comment]
@reynhardtstef Hi dude. I haven't considered creating a repo to upload the TensorRT's SAM2 inference code for the time being. For the conversion of ONNX, I basically refer to this repo [OrtInference](https://github.com/Aimol-l/OrtInference). Only due to project reasons, some places have been optimized. You can refer to this project for conversion.

**Comment by okideal** [comment]
@yformer It works fine. But still the memory attention module costs at most time of whole process (15ms/25ms).

**Comment by huihui6666** [comment]
@okideal > @yformer It works fine. But still the memory attention module costs at most time of whole process (15ms/25ms).

I'm very interested in deploying with TensorRT. Could you share with me whether you deploy multi-object tracking and detection prompts are added during the tracking process?

**Comment by okideal** [comment]
@huihui6666 I think MOT can simply be regarded as multiple replicas of SOT. Just pay attention to the post-processing of matching method. I haven't integrated MOT. Now I just use SAM2 as SOT. As for box prompt, it is applicable to any detector such as YOLO.

**Comment by yformer** [comment]
> [@okideal](https://github.com/okideal) > [@yformer](https://github.com/yformer) It works fine. But still the memory attention module costs at most time of whole process (15ms/25ms).
> 
> I'm very interested in deploying with TensorRT. Could you share with me whether you deploy multi-object tracking and detection prompts are added during the tracking process?

@huihui6666, here is the example notebook for [multi-object tracking](https://github.com/yformer/EfficientTAM/blob/main/notebooks/example_video.ipynb). For detection, you can use detectors to generate box for feeding to our EfficientTAM model.

**Comment by huihui6666** [comment]
> [@yformer](https://github.com/yformer) I have integrated TensorRT in C++. So far, the testing results look good, taking about 20 ms for full process inference on a 4090 GPU (with the efficienttam_s.pt version).
> 
> For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (the dynamic one is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?

Hi,I deploy the model using TensorRT, but I found that the error during tracking increases over time. Have you encountered this situation?

**Comment by okideal** [comment]
@huihui6666 which degradation specifically? Could you post some images for illustration?

**Comment by huihui6666** [comment]
> [@huihui6666](https://github.com/huihui6666) which degradation specifically? Could you post some images for illustration?

@okideal I found that the object is lost after it moves out of the image. I guess it may be caused by some errors in each step. For example, the image encoding part brings a precision error of 3 decimal places. After passing sam_head, the error increases to 2 decimal places. Then this error will increase as the tracking progresses.

**Comment by okideal** [comment]
@huihui6666 Isn't it correct that the target loses tracking when it moves out of the image? Are you referring to the target moving out of the image and then re-entering the image range? If so, there is a probability of both re-lock and losing the target, depending on the complexity of the background & target.

In my experiments, the overall performance of efficientTam is slightly worse than the original SAM2, but it is indeed faster and deployable. I used fp16 precision here and didn't notice any significant accuracy drop; A difference in the second or third decimal place is normal, and deep networks can handle such minor variations.

**Comment by rodneyML** [comment]
> > [@huihui6666](https://github.com/huihui6666) which degradation specifically? Could you post some images for illustration?
> 
> [@okideal](https://github.com/okideal) I found that the object is lost after it moves out of the image. I guess it may be caused by some errors in each step. For example, the image encoding part brings a precision error of 3 decimal places. After passing sam_head, the error increases to 2 decimal places. Then this error will increase as the tracking progresses.

Hi @huihui6666 , make sure your pipeline is correct. I have the following pipeline:

```
Image -> Preprocess -> Image Encoder -> Memory Fusion -> Memory Attention -> Image Decoder (Mask decoder) -> Memory Encoder. 
```

The outputs of the Memory encoder will then be fed into the memory fusion part (note that it might not be the correct term, but that is the part where you manage the memory before giving it to the memory attention model).

In EfficientTAM, refer to [this](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/efficienttam_base.py#L499) function (`_prepare_memory_conditioned_features` - one can call this memory fusion). This part is crucial as it impacts the ability to find the object after it was lost. It is here in this function where @okideal [mentioned](https://github.com/yformer/EfficientTAM/issues/7#issuecomment-2562278198) to pre-fill the queue so that the memory attention model receives static input.

**Comment by okideal** [comment]
@reynhardtstef Dude, have you try some lightweight VOS tracking model which could offer comparable capability as SAM2? EfficientTAM is still kinda of heavy on our embedded platform.

**Comment by rodneyML** [comment]
> [@reynhardtstef](https://github.com/reynhardtstef) Dude, have you try some lightweight VOS tracking model which could offer comparable capability as SAM2? EfficientTAM is still kinda of heavy on our embedded platform.

Not yet no, but I would actually be very interested in that. On what platform are you deploying, I am currently using `torch.compile`, which is also pretty fast (maybe even faster than TensorRT) - but does not have a C++ API.

**Comment by okideal** [comment]
@reynhardtstef As you know, the model has already been successfully run on a GPU using TensorRT. Now, the plan is to deploy this model on an embedded AI chip, where a C++ implementation is necessary. However, deploying the image encoder alone takes 20+ ms per frame, so it’s estimated that the entire pipeline will struggle to achieve real-time performance.

**Comment by huihui6666** [comment]
> > > [@huihui6666 ](https://github.com/huihui6666)具体是哪种退化？能发些图说明一下吗？
> > 
> > 
> > [@okideal ](https://github.com/okideal)我发现物体移出图像后就丢失了。我猜测这可能是由于每个步骤中的一些错误造成的。例如，图像编码部分带来了3位小数的精度误差。经过sam_head之后，误差增加到2位小数。然后，随着跟踪的进行，这个误差会越来越大。
> 
> 你好 [@huihui6666](https://github.com/huihui6666)，请确保您的管道正确。我有以下管道：
> 
> ```
> Image -> Preprocess -> Image Encoder -> Memory Fusion -> Memory Attention -> Image Decoder (Mask decoder) -> Memory Encoder. 
> ```
> 
> 然后，记忆编码器的输出将被输入到记忆融合部分（请注意，这可能不是正确的术语，但这是在将记忆提供给记忆注意模型之前管理记忆的部分）。
> 
> 在 EfficientTAM 中，参考 [此 ](https://github.com/yformer/EfficientTAM/blob/abcd061ebd3cc6e7527d152d75b890126aaa53f6/efficient_track_anything/modeling/efficienttam_base.py#L499)功能（ `_prepare_memory_conditioned_features`——我们可以称之为记忆融合）。这部分至关重要，因为它会影响丢失后找到物品的能力。在这个函数中 [@okideal](https://github.com/okideal) [提到 ](https://github.com/yformer/EfficientTAM/issues/7#issuecomment-2562278198)预填充队列，以便记忆注意力模型接收静态输入。

Thank you, I will check my code carefully.

**Comment by rodneyML** [comment]
> [@reynhardtstef](https://github.com/reynhardtstef) As you know, the model has already been successfully run on a GPU using TensorRT. Now, the plan is to deploy this model on an embedded AI chip, where a C++ implementation is necessary. However, deploying the image encoder alone takes 20+ ms per frame, so it’s estimated that the entire pipeline will struggle to achieve real-time performance.

That sounds interesting, I believe a way to further improve the speed, is to perform int8 calibration on EfficientTAM, but I predict that it will be quite a challenge to succeed in int8 calibration. I know that its possible to make the image encoder and image decoder int8 in OpenVINO ([notebook](https://github.com/openvinotoolkit/openvino_notebooks/blob/latest/notebooks/sam2-image-segmentation/segment-anything-2-image.ipynb)), but I have not yet found a way to get it to work with TensorRT.

**Comment by okideal** [comment]
@reynhardtstef Yes. The AI chip company offers the tool kit for int8/int16 conversion in general, but still there is risk of accuracy degradation on VIT models.

**Comment by huihui6666** [comment]
> [@huihui6666](https://github.com/huihui6666)具体是哪种退化？你能贴一些图片来说明一下吗？
@okideal When I use tensorRT to inference on the memory_attention, I found that it does not support dynamic dimension inference. I used the c++ API and trtexec tool to convert but it didn't work. Have you encountered this problem?

**Comment by okideal** [comment]
> For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (the dynamic one is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?

@huihui6666 My temporary solution above.

**Comment by huihui6666** [comment]
> > For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (the dynamic one is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?
> 
> [@huihui6666](https://github.com/huihui6666) My temporary solution above.

@okideal 
I use TensorRT 10, which supports dynamic dimensions of attention. I have also tried padding the dimensions to the maximum dimension, but the result is wrong, so I think this method is not OK.

**Comment by rodneyML** [comment]
> > > For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (the dynamic one is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?
> > 
> > 
> > [@huihui6666](https://github.com/huihui6666) My temporary solution above.
> 
> [@okideal](https://github.com/okideal) I use TensorRT 10, which supports dynamic dimensions of attention. I have also tried padding the dimensions to the maximum dimension, but the result is wrong, so I think this method is not OK.

I also use the method that @okideal suggested [here](https://github.com/yformer/EfficientTAM/issues/7#issuecomment-2562278198), and it works perfectly fine. The memory attention model requires input from the memory fusion part (`prepare_memory_conditioned_features` function in `efficienttam_base.py`). What I did was to basically just fuse the conditioning frame outputs with copies to form a input that is equal to the max size for the memory attention model. Then all the frames after that simply shifts out the oldest `non_cond_frame_outputs` dict entry in `output_dict`, keeping the fused memory's dimensions the same throughout the whole process.

**Comment by huihui6666** [comment]
> > > > For the deployment of the **memory_attention** model, to make TensorRT use a static graph during conversion (the dynamic one is slow due to operator fusion issues in rope attention), I manually filled the queue with obj_ptr, maskmem_features and maskmem_pos_enc_ while initializing the anchor boxes for tracking. I would like to know if such an operation theoretically affects the output results?
> > > 
> > > 
> > > [@huihui6666](https://github.com/huihui6666) My temporary solution above.
> > 
> > 
> > [@okideal](https://github.com/okideal) I use TensorRT 10, which supports dynamic dimensions of attention. I have also tried padding the dimensions to the maximum dimension, but the result is wrong, so I think this method is not OK.
> 
> I also use the method that [@okideal](https://github.com/okideal) suggested [here](https://github.com/yformer/EfficientTAM/issues/7#issuecomment-2562278198), and it works perfectly fine. The memory attention model requires input from the memory fusion part (`prepare_memory_conditioned_features` function in `efficienttam_base.py`). What I did was to basically just fuse the conditioning frame outputs with copies to form a input that is equal to the max size for the memory attention model. Then all the frames after that simply shifts out the oldest `non_cond_frame_outputs` dict entry in `output_dict`, keeping the fused memory's dimensions the same throughout the whole process.

@reynhardtstef  Thanks very much.Your suggestion work well!
